### ISSUE TITLE
S3 method Smith dot not exported, only the generic

### ISSUE THREAD

**Issue by rkrug** [klutometis/roxygen#805]
Hi
I have the following definitions:
```
#' @export
#'
as.plantuml <- function(x, ...) {
  UseMethod("as.plantuml", x)
}
```

and

```
#' @export
#'
as.plantuml.list <- function(
  x,
  complete = FALSE,
  nm = NULL,
  ...
) {
  puml <- "
  return(puml)
}
```

which results in a `NAMESPACE` file as followed:
```
# Generated by roxygen2: do not edit by hand

S3method(as.plantuml,list)
export(as.plantuml)
```

So the function `as.plantuml.list()` is not exported and only accessible via `plantuml:::as.plantuml.list()`. I would like too have it exported, to make it easier to define new generics in another package.

Is this a bug for am I doing something wrong?

Please see [https://github.com/rkrug/plantuml](https://github.com/rkrug/plantuml) for the complete code.

Thanks.

**Comment by rkrug** [comment]
Unfortunately, this makes it slightly more cumbersome, as the definition of the new generic now looks as followed:

```
as.plantuml.md_dataCite <- function(  
  x,
  complete = FALSE,
  nm = NULL,
  ...
) {
  plantuml:::as.plantuml.list( x = x, complete = complete, nm = nm )
}
```

So I effectively have to live with using `:::` in this case?

**Comment by gaborcsardi** [comment]
I am not sure what you are trying to do, but normally you don't called the methods directly. You call the generic, and  the method is selected  automatically  based on the type of `x`.  E.g. `plantuml:::as.plantuml(x, ...)` will call `plantuml:::as.plantuml.list()` if  `x` is a list.

**Comment by gaborcsardi** [comment]
Here is example from  the utils package, which defines and exports the `str` generic. It also defines the `str.data.frame` method, but it does not export it:
```r
❯ str
function (object, ...)
UseMethod("str")
<environment: namespace:utils>

❯ str.data.frame
Error: object 'str.data.frame' not found

❯ str(iris)
'data.frame':	150 obs. of  5 variables:
 $ Sepal.Length: num  5.1 4.9 4.7 4.6 5 5.4 4.6 5 4.4 4.9 ...
 $ Sepal.Width : num  3.5 3 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 ...
 $ Petal.Length: num  1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4 1.5 ...
 $ Petal.Width : num  0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 ...
 $ Species     : Factor w/ 3 levels "setosa","versicolor",..: 1 1 1 1 1 1 1 1 1 1 ...
```

**Comment by rkrug** [comment]
Sure - the generic is called automatically based on the type. But the object `x` is not of class `list` (although it can be treated as one) and I want to treat it as a list only in the context of the `as.plantuml` function.

I could simply add `list` to the class in the constructor (my code - I am in control), but I would like to simply say that for `as.plantuml` the class `md_dataCite` should be treated ads a `list`. 

so I think that the code using `plantuml:::as.plantuml.list()`` is my only option?

**Comment by gaborcsardi** [comment]
> Sure - the generic is called automatically based on the type

No, the generic is `as.plantuml`, you need to call that directly. 

Maybe you just want to call `as.plantuml.list()` from `as.plantuml.md_dataCite()`, i,e. like calling the parent method in a "regular"  OO system? In S3 this is done with `NextMethod()`:

```r
as.plantuml.md_dataCite <- function(  
  x,
  complete = FALSE,
  nm = NULL,
  ...
) {
  NextMethod("as.plantuml")
}

```

Here is a complete example from the Advanced R book (http://adv-r.had.co.nz/S3.html):
```r
#' @export
baz <- function(x) UseMethod("baz", x)

#'  @export
baz.A <- function(x) "A"

#' @export
baz.B <- function(x) "B"

#' @export
baz.C <- function(x) c("C", NextMethod())
ca <- structure(1, class = c("C", "A"))
cb <- structure(1, class = c("C", "B"))
baz(ca)
baz(cb)
```
This will generate the NAMESPACE:
```
export(baz)
S3method(baz,A)
S3method(baz,B)
S3method(baz,C)
```

Anyway, my points are 
-  you don't need to export a method, it is an antipattern to do that.
-  you _can_ export the method if you really want to. Just say `#' @export as.plantuml.list`. But THIS might just get you into trouble later, because it is not the R idiom.
- this does not have much to do with roxygen2, it is how base R works.
- hundreds of packages work the same way, maybe you want to look at some examples: https://github.com/search?q=user%3Acran+NextMethod&type=Code
